File: guindo/TP_bus_track/backend/models/ligne.py
from dataclasses import dataclass, field
from typing import List
from models.arret import Arret


@dataclass
class Ligne:
    """Représente une ligne de bus avec ses arrêts."""
    id: int
    nom: str
    couleur: str
    description: str
    arrets: List[Arret] = field(default_factory=list)

    def get_arret(self, index: int) -> Arret | None:
        """Retourne l'arrêt à l'index donné, ou None si hors limites."""
        if 0 <= index < len(self.arrets):
            return self.arrets[index]
        return None

    def nombre_arrets(self) -> int:
        return len(self.arrets)

    def __repr__(self):
        return f"Ligne(id={self.id}, nom='{self.nom}', arrets={self.nombre_arrets()})"

File: Projet Martibus/backend/models/Ligne.js
/**
 * models/Ligne.js
 * Requêtes SQL liées aux lignes de bus.
 * Fonctionne en mode PostgreSQL ou en mode mémoire (fallback).
 */

const { pool, memoryDB, isMemory } = require('../config/db');

const Ligne = {
  /** Retourne toutes les lignes actives. */
  findAll: async () => {
    if (isMemory()) return memoryDB.lignes.filter((l) => l.actif);
    const { rows } = await pool.query(`
      SELECT id, nom, couleur, description, actif
      FROM lignes
      WHERE actif = TRUE
      ORDER BY nom
    `);
    return rows;
  },

  /**
   * Retourne une ligne avec ses arrêts dans l'ordre.
   * @param {number} id
   */
  findById: async (id) => {
    if (isMemory()) return memoryDB.getLigneWithArrets(id);

    // Données de la ligne
    const { rows: ligneRows } = await pool.query(`
      SELECT id, nom, couleur, description, actif
      FROM lignes
      WHERE id = $1
    `, [id]);

    if (!ligneRows[0]) return null;

    // Arrêts de la ligne, dans l'ordre
    const { rows: arretRows } = await pool.query(`
      SELECT
        a.id,
        a.nom,
        a.latitude,
        a.longitude,
        la.ordre
      FROM arrets a
      JOIN ligne_arrets la ON la.arret_id = a.id
      WHERE la.ligne_id = $1
      ORDER BY la.ordre
    `, [id]);

    return { ...ligneRows[0], arrets: arretRows };
  },
};

module.exports = Ligne;
